File: Front-End/HMS_Front/src/services/testRequestService.ts
import api from '@/lib/api';

export interface TestRequestDto {
  id: number;
  testType: string;
  description?: string;
  priority: string;
  status: string;
  reportUrl?: string;
  results?: string;
  charges?: number;
  requestedAt: string;
  completedAt?: string;
  patientId: number;
  patientName: string;
  doctorId: number;
  doctorName: string;
  technicianId?: number;
  technicianName?: string;
}

export interface CreateTestRequestPayload {
  testType: string;
  description?: string;
  priority?: string;
  patientId: number;
  doctorId: number;
  charges?: number;
}

/** Normalize both PageImpl (old) and PagedModel VIA_DTO (new) response shapes */
const mapPage = (data: any): { content: TestRequestDto[]; totalElements: number } => ({
  content: data.content ?? [],
  totalElements: data.page?.totalElements ?? data.totalElements ?? 0,
});

const testRequestService = {
  async getAll(): Promise<TestRequestDto[]> {
    const { data } = await api.get<TestRequestDto[]>('/api/test-requests');
    return data;
  },

  async getById(id: number | string): Promise<TestRequestDto> {
    const { data } = await api.get<TestRequestDto>(`/api/test-requests/${id}`);
    return data;
  },

  async getByPatient(patientId: number | string): Promise<TestRequestDto[]> {
    const { data } = await api.get<TestRequestDto[]>(`/api/test-requests/patient/${patientId}`);
    return data;
  },

  async getByDoctor(doctorId: number | string): Promise<TestRequestDto[]> {
    const { data } = await api.get<TestRequestDto[]>(`/api/test-requests/doctor/${doctorId}`);
    return data;
  },

  async getByTechnician(technicianId: number | string, page = 0, size = 50): Promise<{ content: TestRequestDto[]; totalElements: number }> {
    const { data } = await api.get(`/api/test-requests/technician/${technicianId}`, { params: { page, size } });
    return mapPage(data);
  },

  async getByStatus(status: string): Promise<TestRequestDto[]> {
    const { data } = await api.get<TestRequestDto[]>(`/api/test-requests/status/${status}`);
    return data;
  },

  async create(payload: CreateTestRequestPayload): Promise<TestRequestDto> {
    const { data } = await api.post<TestRequestDto>('/api/test-requests', payload);
    return data;
  },

  async assignTechnician(id: number | string, technicianId: number | string): Promise<TestRequestDto> {
    const { data } = await api.put<TestRequestDto>(`/api/test-requests/${id}/assign/${technicianId}`);
    return data;
  },

  async acknowledge(id: number | string): Promise<TestRequestDto> {
    const { data } = await api.put<TestRequestDto>(`/api/test-requests/${id}/acknowledge`);
    return data;
  },

  async start(id: number | string): Promise<TestRequestDto> {
    const { data } = await api.put<TestRequestDto>(`/api/test-requests/${id}/start`);
    return data;
  },

  async complete(id: number | string, results: string, reportUrl: string, charges?: number): Promise<TestRequestDto> {
    const { data } = await api.put<TestRequestDto>(`/api/test-requests/${id}/complete`, { results, reportUrl, charges });
    return data;
  },

  async uploadReport(id: number | string, file: File): Promise<{ reportUrl: string }> {
    const formData = new FormData();
    formData.append('file', file);
    const { data } = await api.post<{ reportUrl: string }>(`/api/test-requests/${id}/upload-report`, formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
    });
    return data;
  },

  async cancel(id: number | string): Promise<TestRequestDto> {
    const { data } = await api.put<TestRequestDto>(`/api/test-requests/${id}/cancel`);
    return data;
  },

  async delete(id: number | string): Promise<void> {
    await api.delete(`/api/test-requests/${id}`);
  },
};

export default testRequestService;

File: Back-End/Back-End/hospital-management-system-main/src/main/java/com/hospital/hms/service/TestRequestService.java
package com.hospital.hms.service;

import com.hospital.hms.dto.TestRequestDTO;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

public interface TestRequestService {

    /** Doctor creates a test request */
    TestRequestDTO createTestRequest(TestRequestDTO dto);

    List<TestRequestDTO> getAll();

    TestRequestDTO getById(Long id);

    List<TestRequestDTO> getByPatient(Long patientId);

    List<TestRequestDTO> getByDoctor(Long doctorId);

    /** Technician views their assigned tests */
    Page<TestRequestDTO> getByTechnician(Long technicianId, Pageable pageable);

    List<TestRequestDTO> getByStatus(String status);

    /** Admin/Doctor assigns a test to a technician */
    TestRequestDTO assignTechnician(Long testRequestId, Long technicianId);

    /** Technician acknowledges the test */
    TestRequestDTO acknowledge(Long testRequestId);

    /** Technician marks test as in-progress */
    TestRequestDTO startProcessing(Long testRequestId);

    /**
     * Technician completes the test:
     * - sets status = COMPLETED
     * - stores results & reportUrl
     * - triggers notification to doctor
     * - triggers billing (creates invoice item)
     */
    TestRequestDTO complete(Long testRequestId, String results, String reportUrl, Double charges);

    /** Technician uploads a report file and returns the stored path/URL */
    String uploadReport(Long testRequestId, MultipartFile file);

    TestRequestDTO cancel(Long testRequestId);

    void deleteTestRequest(Long id);
}
